neuralmonkey/decoders/decoder.py

```diff
@@ -1,415 +1,357 @@
+            encoders: List of encoders whose outputs will be decoded
+            vocabulary: Output vocabulary
+            data_id: Identifier of the data series fed to this decoder
+
+        Keyword arguments:
+            embedding_size: Size of embedding vectors. Default 200
+            max_output: Maximum length of the output. Default 20
+            rnn_size: When projection is used or when no encoder is supplied,
+                      this is the size of the projected vector.
+            dropout_keep_prob: Dropout keep probability. Default 1 (no dropout)
+
+        Flags:
+            use_attention: Indicates whether to use attention from encoders
+            reuse_word_embeddings: Boolean flag specifying whether to reuse
+                                   word embeddings. If True, word embeddings
+                                   from the first encoder will be used
+            project_encoder_outputs: Boolean flag whether to project output
+                                     states of encoders
 
-        """A class that collects the part of the computation graph that is
-        needed for decoding.
+        """
+        self.encoders = encoders
+        self.vocabulary = vocabulary
+        self.data_id = data_id
 
-        TensorBoard summaries are collected in this class into the following
-        collections:
+        self.max_output = kwargs.get("max_output", 20)
+        self.embedding_size = kwargs.get("embedding_size", 200)
+        self.name = kwargs.get("name", "decoder")
+        dropout_keep_prob = kwargs.get("dropout_keep_prob", 1.0)
 
-            * 'summary_train' - collects statistics from the train-time
+        self.use_attention = kwargs.get("use_attention", False)
+        self.reuse_word_embeddings = kwargs.get("reuse_word_embeddings", False)
+        self.project_encoder_outputs = kwargs.get("project_encoder_outputs",
+                                                  False)
 
-            * 'sumarry_val' - collects OAstatistics while being tested on the
-                 development data
+        log("Initializing decoder, name: '{}'".format(self.name))
 
-        Arguments:
+        if self.project_encoder_outputs or len(self.encoders) == 0:
+            self.rnn_size = kwargs.get("rnn_size", 200)
+        else:
+            self.rnn_size = sum([e.encoded.get_shape()[1].value
+                                 for e in self.encoders])
 
-            encoders: List of encoders. If no encoder is provided, the decoder
-                can be used to train a LM.
+        ### Initialize model
 
-            vocabulary: Vocabulary used for decoding
+        self.dropout_placeholder = tf.placeholder_with_default(
+            tf.constant(dropout_keep_prob, tf.float32),
+            shape=[], name="decoder_dropout_placeholder")
 
-            data_id:
+        state = self.initial_state()
 
-            rnn_size: Size of the RNN state.
+        self.weights, self.biases = self.state_to_output()
```

Comment:
Měl jsem tady rozepsanej koment, ale někam asi zapadnul.. Opravil jsem to, ať jdem na ruku puritánům.